### PR TITLE
chore: bump typescript-eslint to v8

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
👋 Hi! Coming over from https://github.com/typescript-eslint/typescript-eslint/issues/9141: we're working on typescript-eslint v8 and would like to try the beta out on repos like this one.

I'm sending this draft PR as a reference to see what issues we might give you in upgrading and to try to be helpful. If this is annoying noise to you, please forgive me, I'll withdraw the PR. But I hope this is useful on your end - and please let me know if you have any feedback! ❤️ 

I've put inline comments on any changes. There shouldn't be anything user-facing (this isn't a `feat:` or `fix:` PR).

For reference, #1476 was a similar PR that bumped typescript-eslint to v6. Hello again! 😄

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -12,12 +12,7 @@ const config = {
     ecmaVersion: "latest",
     sourceType: "module",
     tsconfigRootDir: __dirname,
-    project: [
-      "./tsconfig.json",
-      "./cli/tsconfig.eslint.json", // separate eslint config for the CLI since we want to lint and typecheck differently due to template files
-      "./upgrade/tsconfig.json",
-      "./www/tsconfig.json",
-    ],
+    projectService: true,
```
Comment (JoshuaKGoldberg) [inline .eslintrc.cjs:15]: https://typescript-eslint.io/blog/announcing-typescript-eslint-v8-beta#project-service: this is our snazzy new way of enabling typed linting. It should be faster and require less configuration.

I couldn't tell from quickly reading the code: what is the `tsconfig.eslint.json` doing differently? Is it still necessary?

---

Btw, I ran [hyperfine](https://github.com/sharkdp/hyperfine) measurements, and they showed promising performance improvements on my M1 Max Mac Studio:

* ~10% faster bumping to typescript-eslint from v7 to v8
* ~20% faster again switching from `project` to `projectService`

```plaintext
Benchmark 1: npx turbo lint --no-cache
  Time (mean ± σ):      7.464 s ±  0.094 s    [User: 32.604 s, System: 2.431 s]
  Range (min … max):    7.346 s …  7.665 s    10 runs
```

```plaintext
# baseline: typescript-eslint@v7 with project
Benchmark 1: npx turbo lint --no-cache
  Time (mean ± σ):      7.464 s ±  0.094 s    [User: 32.604 s, System: 2.431 s]
  Range (min … max):    7.346 s …  7.665 s    10 runs
```
```plaintext
# typescript-eslint@v8 with project
Benchmark 1: npx turbo lint --no-cache
  Time (mean ± σ):      6.662 s ±  0.235 s    [User: 29.944 s, System: 2.253 s]
  Range (min … max):    6.501 s …  7.254 s    10 runs
```

```plaintext
# typescript-eslint@v8 with projectService
Benchmark 1: npx turbo lint --no-cache
  Time (mean ± σ):      5.344 s ±  0.126 s    [User: 25.637 s, System: 1.683 s]
  Range (min … max):    5.204 s …  5.554 s    10 runs
```

...for an overall performance improvement of roughly **28%**.

File: cli/src/index.ts [hunk 2]
```diff
@@ -31,7 +31,9 @@ const main = async () => {
   const npmVersion = await getNpmVersion();
   const pkgManager = getUserPkgManager();
   renderTitle();
-  npmVersion && renderVersionWarning(npmVersion);
+  if (npmVersion) {
+    renderVersionWarning(npmVersion);
+  }
```
Comment (JoshuaKGoldberg) [inline cli/src/index.ts:36]: [`@typescript-eslint/no-unused-expressions`](https://typescript-eslint.io/rules/no-unused-expressions) was promoted to the [`recommended` preset](https://v8--typescript-eslint.netlify.app/users/configs#recommended) in v8.

File: cli/src/helpers/logNextSteps.ts [hunk 3]
```diff
@@ -23,7 +23,9 @@ export const logNextSteps = async ({
   const pkgManager = getUserPkgManager();
 
   logger.info("Next steps:");
-  projectName !== "." && logger.info(`  cd ${projectName}`);
+  if (projectName !== ".") {
+    logger.info(`  cd ${projectName}`);
+  }
```
Comment (JoshuaKGoldberg) [inline cli/src/helpers/logNextSteps.ts:28]: [`@typescript-eslint/no-unused-expressions`](https://typescript-eslint.io/rules/no-unused-expressions) was promoted to the [`recommended` preset](https://v8--typescript-eslint.netlify.app/users/configs#recommended) in v8.

File: cli/src/utils/renderVersionWarning.ts [hunk 4]
```diff
@@ -55,12 +55,14 @@ function checkForLatestVersion(): Promise<string> {
...
           } else {
+            // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
             reject();
           }
         }
       )
       .on("error", () => {
         // logger.error("Unable to check for latest version.");
+        // eslint-disable-next-line @typescript-eslint/prefer-promise-reject-errors
         reject();
```
Comment (JoshuaKGoldberg) [inline cli/src/utils/renderVersionWarning.ts:66]: This is the first time I can remember seeing `reject()` called with no argument. Interesting!

[`@typescript-eslint/prefer-promise-reject-errors`](https://typescript-eslint.io/rules/prefer-promise-reject-errors) was promoted to the [`strict-type-checked` config](https://v8--typescript-eslint.netlify.app/users/configs/#strict-type-checked) in v8.

File: upgrade/src/app/api/og/route.tsx [hunk 5]
```diff
@@ -117,8 +117,8 @@ async function getFont<TWeights extends readonly number[]>({
     .split("@font-face {")
     .splice(1)
     .map((font) => {
-      const u = font.match(/src: url\((.+)\) format\('(opentype|truetype)'\)/);
-      const w = font.match(/font-weight: (\d+)/);
+      const u = /src: url\((.+)\) format\('(opentype|truetype)'\)/.exec(font);
+      const w = /font-weight: (\d+)/.exec(font);
```
Comment (JoshuaKGoldberg) [inline upgrade/src/app/api/og/route.tsx:121]: [`@typescript-eslint/prefer-regexp-exec`](https://v8--typescript-eslint.netlify.app/rules/prefer-regexp-exec) was promoted to the [`stylistic-type-checked` config](https://v8--typescript-eslint.netlify.app/users/configs/#stylistic-type-checked) in v8.

File: www/src/languages.ts [hunk 6]
```diff
@@ -32,7 +32,7 @@ export function getIsRtlFromUrl(pathname: string) {
 }
 
 export function getIsRtlFromLangCode(language: KnownLanguageCode) {
-  if (rtlLanguages.indexOf(language) !== -1) {
+  if (rtlLanguages.includes(language)) {
```
Comment (JoshuaKGoldberg) [inline www/src/languages.ts:35]: [`@typescript-eslint/prefer-includes`](https://v8--typescript-eslint.netlify.app/rules/prefer-includes) was promoted to the [`stylistic-type-checked` config](https://v8--typescript-eslint.netlify.app/users/configs/#stylistic-type-checked) in v8.

File: .eslintrc.cjs (hunk 1 above)
Comment (juliusmarminge) [reply .eslintrc.cjs:15]: > I couldn't tell from quickly reading the code: what is the tsconfig.eslint.json doing differently? Is it still necessary?

I can't remember why it's there but from guessing it sounds like `projectService` does it for us :)

File: cli/src/utils/renderVersionWarning.ts (hunk 4 above)
Comment (juliusmarminge) [reply cli/src/utils/renderVersionWarning.ts:66]: probably should reject with a descriptive error xd

File: cli/template/base/package.json [hunk 7]
```diff
@@ -22,8 +22,8 @@
     "@types/node": "^20.11.20",
     "@types/react": "^18.2.57",
     "@types/react-dom": "^18.2.19",
-    "@typescript-eslint/eslint-plugin": "^7.1.1",
-    "@typescript-eslint/parser": "^7.1.1",
+    "@typescript-eslint/eslint-plugin": "8.0.0-alpha.39",
+    "@typescript-eslint/parser": "8.0.0-alpha.39",
```
Comment (juliusmarminge) [inline cli/template/base/package.json:26]: this is user-facing so we'll hold off until it's stable I tihnk

Comment (JoshuaKGoldberg) [reply cli/template/base/package.json:26]: Oh right I'd meant to come back to this, sorry 😅 doing now!

**Comment by JoshuaKGoldberg** [comment]
I don't know what's going on with CI, but 🤷 putting in review.

**Comment by juliusmarminge** [comment]
This looks like a version mismatch in next-auth's types...

CleanShot 2024-08-19 at 10 08 19

File: package.json [hunk 8]
```diff
@@ -63,5 +63,9 @@
     "prettier": "^3.3.2",
     "turbo": "1.13.3-canary.3",
     "typescript": "^5.5.3"
+  },
+  "resolutions": {
+    "@typescript-eslint/eslint-plugin": "8.1.0",
+    "@typescript-eslint/parser": "8.1.0"
```
Comment (juliusmarminge) [inline package.json:69]: can this be removed nnow?

**Review by juliusmarminge — APPROVED** [review]

File: package.json (hunk 8 above)
Comment (juliusmarminge) [reply package.json:69]: looks like something resolved to v7 when removing this so leaving it in

**Comment by JoshuaKGoldberg** [comment]
Awesome, thanks @juliusmarminge! ❤️

**Pull request merged**